Save each note in a separate file and include saving labels and attachments

Zettt opened the pull request:
See readme for more detail.

Zettt committed:
Remove the self-advertisement.

Zettt committed:
Include more metadata and general improvements

I'm sorry I couldn't break these changes down better. We're now saving attachments properly, include the labels, and have generally better logging (maybe a bit too verbose).

Zettt committed:
Update README.md

tpwo commented on README.md:
```diff
@@ -22,6 +21,14 @@ Copy `Takeout` folder into root of this repo, and run the script:
-separated with `---`.
+After running the script, Markdown files with timestamp in the filename are created and written to an `export` folder.
+
+## Main changes from forked repo:
```
```suggestion
## Main features
```

tpwo commented on google_keep_extractor.py:
```diff
@@ -22,25 +24,24 @@
-    print(f'File {EXPORT_PATH} saved.')
 
 
 def _get_md_header() -> str:
```
This function is now useless. I guess that if we have an md file per note, metadata about export time shouldn't be in each file.

I think we can just remove it.

tpwo commented on google_keep_extractor.py:
```diff
@@ -51,6 +52,8 @@ def _load_notes(folder: pathlib.Path) -> list[Note]:
                 notes.append(_load_note(item))
             except RuntimeError as err:
                 print(f"Skipping note: '{err}'")
+            except Exception as e:
```
Maybe we can leave only this generic except with `Exception`? I don't think a separate one for RuntimeError is needed.

Also, could you use `err` for msg variable? I try to avoid using single letter names even in such a short context

Zettt replied:
Makes sense. I'm not exactly sure anymore why I needed this. I ran into some sort of error when processing the JSON files and this resolved the issue. What do you suggest? Keep it in or delete?

tpwo replied:
I think we can go with `Except`. I noticed you only added new lines of code after my comments, and haven't removed the old ones. No worries, I fixed it :)

tpwo commented on google_keep_extractor.py:
```diff
@@ -90,13 +98,12 @@ def _get_title(note: dict[str, object]) -> str:
-            return text
-    except KeyError:
+        return text
+    elif 'listContent' in note:
```
Could you create a CONSTANT e.g. `JSON_NOTE_LIST` and use it for each occurence of this string? 

I see I haven't done it for my existing code, so we can clean it up now.

tpwo commented on google_keep_extractor.py:
```diff
@@ -108,11 +115,41 @@ def _get_text(note: dict[str, object]) -> str:
+    )
+    labels_str = ', '.join(note.labels)
+    labels_line = f'\n\nLabels: {labels_str}' if labels_str else ''
+    return f'## {note.title}\n\n{note.text}\n\n{attachments_str}{labels_line}\n\n---\n\n'
```
I think `---` is not needed if we only have a single not per file, right?

Zettt replied:
So the labels are the ones created on Google Keep. These are like tags or categories. It makes sense to keep them, at least I wanted to. But you are right in the sense that we shouldn't write an otherwise empty "Labels:" line. I will remove it. I hope I can work on it today or tomorrow.

tpwo replied:
Makes sense, I see that it works correctly at the moment.

But I meant to remove the trailing whitespace and `---` at the end of each note. I used it before to separate notes from each other back when everything was in a single file

You can see it here: 
![image](https://github.com/user-attachments/assets/688405d3-5942-445f-a4d9-f9d4fce5a55e)

tpwo reviewed (commented):
Hey, thanks for the PR. Changes look good, I like the new functionality. I added some comments, and it would be nice to address them before merging, mostly small things :)

Zettt committed:
Removing dead code

Zettt committed:
Introduce JSON_NOTE_LIST constant

Zettt committed:
Suppress labels_line when a notes hasn't any labels

Zettt committed:
Throw an error for execeptions too.

Zettt commented:
I hope I resolved everything correctly. My first pull request that requested changes after my first submit.

tpwo committed:
Remove duplicated lines

tpwo committed:
Leave out only one except block

tpwo committed:
Enchance `_note_to_str`, so it don't spawn more white-space than needed

Include a docstring for it.

tpwo committed:
Save attachments to `attachments` folder

tpwo committed:
Update EXPORT_PATH to be just the `export` folder

tpwo committed:
Update gitignore

tpwo committed:
Use backticks in printed output for dynamically generated strings

This way, it's easier to analyze the output.

tpwo committed:
Redesign the note filename creation process

Add `created_at` field to the Note dataclass. If title is empty, this
field is used to serve as a note and filename title. Otherwise the title
and timestamp are concacinated.

Regex sanitizing title is updated to remove illegal characters instead
replacing them with `_`, and each space is replaced with `_`, as
white-space in file names is annoying to handle.

More code in the whole file is updated to correctly handle having title
and timestamp as separate entities.

tpwo committed:
Update readme

tpwo commented:
Sure, thanks for that. I tried running it on my side, but I noticed some issues.

Instead of writing comments I spent some time to fix them. You can see this at the top of the commits I pushed above.

Later, I updated the code to solve other issues I noticed.

1. When labels or attachments were empty, there was a lot of white-space in the generated Markdown files. The new solution takes care of it, so we always have a single blank line between note elements. The only exception is more than one blank line in the note text itself. But I figure it's safer not to touch text at all here, and sometimes more whitespace is intentional.
2. All attachments are now saved to a subfolder, so they're not at the same level as the notes.
3. Generation of Markdown filenames was more complex than needed. I added `created_at` field to Note dataclass, so it can be solved in a better way. I updated your regex to strip illegal characters instead of replacing them with `_`, but now each space is replaced with `_` (perosonally I don't like spaces in file names, they are annoying to handle .
4. Other minor things I noticed along the way.

I'm happy with the final result, it works fine when testing it with my Keep export, so I'm merging the PR. If you have any more improvements feel free to raise another PR :)

Zettt commented:
Ah, sorry. I didn't see that it pushed duplicate lines. I've used VSCode and it shows me the comments from here inline. When I worked on the code I thought I edited everything properly. The tutorial I've used didn't mention there would be a problem. Thanks for taking care of it!
